File: RetrofitRxJavaDagger2Practice/app/src/main/java/com/calvert/retrofitrxjavadagger2practice/MainActivity.java
package com.calvert.retrofitrxjavadagger2practice;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

import com.calvert.retrofitrxjavadagger2practice.model.UsersResponse;
import com.calvert.retrofitrxjavadagger2practice.network.ApiInterface;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();

    @Inject
    Retrofit retrofit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //init component
        ((MyApplication)getApplication()).getComponent().injectRetrofit(this);

        ApiInterface apiInterface = retrofit.create(ApiInterface.class);

        //call endpoint method of observable
        Observable<List<UsersResponse>> userObservable = apiInterface.getUsers();
        //subcribe with two methods using lamda
        //scheduler for background thread
        //mainthread for ui
        userObservable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::handleResults, this::handleError);
    }

    private void handleResults(List<UsersResponse> usersResponse) {
        Log.d(TAG, "handleResults: " + usersResponse.toString());
    }

    private void handleError(Throwable throwable) {
        Log.d(TAG, "handleError: " + throwable.getMessage());
    }
}

File: WeatherKotlinRetrofitRxJavaDagger/app/src/main/java/com/calvert/weatherkotlinretrofitrxjavadagger/MainActivity.kt
package com.calvert.weatherkotlinretrofitrxjavadagger

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.widget.Adapter
import com.calvert.weatherkotlinretrofitrxjavadagger.adapter.WeatherAdapter
import com.calvert.weatherkotlinretrofitrxjavadagger.di.DaggerMyComponent
import com.calvert.weatherkotlinretrofitrxjavadagger.di.MyComponent
import com.calvert.weatherkotlinretrofitrxjavadagger.di.RetrofitModule
import com.calvert.weatherkotlinretrofitrxjavadagger.model.LondonWeather
import com.calvert.weatherkotlinretrofitrxjavadagger.network.ApiInterface
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Retrofit
import javax.inject.Inject

class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var apiInterface: ApiInterface
    lateinit var myComponent: MyComponent

    lateinit var weatherAdapter: WeatherAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        init()

        myComponent = DaggerMyComponent.builder().retrofitModule(RetrofitModule).build()
        myComponent.injectRetrofit(this@MainActivity)
        apiInterface.getLondonWeather()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ it ->
                Log.d("Response", it.weather.toString())
                weatherAdapter = WeatherAdapter(this@MainActivity, it.weather)
                weatherRecyclerView.adapter = weatherAdapter
                weatherAdapter.notifyDataSetChanged()
            }, { it ->
                Log.e("Error", it.message)
            })

//        getWeather()
    }

//    private fun getWeather() {
//        val apiInterface = ApiInterface.getRetrofit().getLondonWeather()
//            .subscribeOn(Schedulers.io())
//            .observeOn(AndroidSchedulers.mainThread())
//            .subscribe({it ->
//                Log.d("Response", it.weather.toString())
//                weatherAdapter = WeatherAdapter(this@MainActivity, it.weather)
//                weatherRecyclerView.adapter = weatherAdapter
//                weatherAdapter.notifyDataSetChanged()
//            }, {it ->
//                Log.e("Error",it.message)
//            })
//    }

    private fun init() {
        weatherRecyclerView.layoutManager = LinearLayoutManager(this)
    }
}
